projectroles/tests/test_permissions.py

```diff
@@ -40,6 +40,11 @@
 class TestPermissionMixin:
     """Helper class for permission tests"""
 
+    # Special attribute for dynamic URL changing
+    # for use-cases where the URL depends on the object
+    # that is being changed in cleanup_method.
+    dynamic_url = None
```

Comment:
Couldn't we just do the following:

- Update an URL variable (e.g. `self.url`) in `cleanup_method`
- Pass that URL to the `assert_*()` method(s) as the `url` argument, whenever we call the method in a loop

That would be more clear, as we would not be dependent on the existence of a specifically named class member. variable Looking at the implementation, this should work just fine, as we now disregard the `url` argument completely if this overriding URL is set. Unless I'm missing something?

Reply:
Good notice! I've implemented that, but also found unwanted behaviour (which can not handle in a simple way):
If: 
- user provides `cleanup_method`, that does not change `url` + 
- there is reference `self.url` in that test + 
- user sets url param in `assert_response*` to some variable different from `self.url` 
- == we get bug :)

However, there's no logic in such a "developer experience", but any vulnerabilities are not welcome ((